File: fabric-audit-agent/core/lifecycle.js
export const ACTIVE_STATES = new Set(['open', 'acknowledged']);
export const DEFAULT_LIFECYCLE = { state: 'open', since: null, snoozeUntil: null, note: null };

/**
 * Split findings into active vs suppressed based on persisted lifecycle states.
 * Snoozed findings whose snoozeUntil has passed are reactivated (open). Pure.
 * @param {object[]} findings  findings with a `.key`
 * @param {Record<string, object>} states  key -> { state, since, snoozeUntil, note }
 * @param {number} nowMs  current time in ms (for snooze expiry); 0 disables expiry
 * @returns {{ active: object[], suppressed: object[] }}
 */
export function applyLifecycle(findings, states = {}, nowMs = 0) {
  const active = [];
  const suppressed = [];
  for (const f of findings) {
    const raw = (f.key && states[f.key]) ? states[f.key] : DEFAULT_LIFECYCLE;
    let lc = { ...DEFAULT_LIFECYCLE, ...raw };
    if (nowMs > 0 && lc.state === 'snoozed' && lc.snoozeUntil != null && Date.parse(lc.snoozeUntil) < nowMs) {
      lc = { ...lc, state: 'open', snoozeUntil: null };
    }
    const annotated = { ...f, lifecycle: lc };
    (ACTIVE_STATES.has(lc.state) ? active : suppressed).push(annotated);
  }
  return { active, suppressed };
}

/**
 * Pure state transition — returns a NEW states map (does not mutate input).
 * @param {Record<string,object>} states
 * @param {string} key
 * @param {'open'|'acknowledged'|'snoozed'|'resolved'|'wontfix'} state
 * @param {{ snoozeUntil?:string, note?:string, now?:string }} [opts]
 */
export function setState(states = {}, key, state, opts = {}) {
  return {
    ...states,
    [key]: {
      state,
      since: opts.now ?? null,
      snoozeUntil: opts.snoozeUntil ?? null,
      note: opts.note ?? null,
    },
  };
}

File: fabric-audit-agent-app/fabric_audit_agent/lifecycle.py
"""Finding lifecycle. Faithful port of ``core/lifecycle.js``.

Split findings into active vs suppressed by persisted state; expired snoozes reactivate.
"""
from datetime import datetime, timezone

ACTIVE_STATES = {"open", "acknowledged"}
DEFAULT_LIFECYCLE = {"state": "open", "since": None, "snoozeUntil": None, "note": None}


def _parse_ms(s):
    """Parse an ISO timestamp to epoch-ms (mirrors JS Date.parse); None if unparseable."""
    if not isinstance(s, str):
        return None
    try:
        dt = datetime.fromisoformat(s.replace("Z", "+00:00"))
    except ValueError:
        return None
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.timestamp() * 1000


def apply_lifecycle(findings, states=None, now_ms=0):
    states = states or {}
    active, suppressed = [], []
    for f in findings:
        raw = states[f["key"]] if (f.get("key") and f["key"] in states) else DEFAULT_LIFECYCLE
        lc = {**DEFAULT_LIFECYCLE, **raw}
        if now_ms > 0 and lc["state"] == "snoozed" and lc.get("snoozeUntil") is not None:
            ms = _parse_ms(lc["snoozeUntil"])
            if ms is not None and ms < now_ms:
                lc = {**lc, "state": "open", "snoozeUntil": None}
        annotated = {**f, "lifecycle": lc}
        (active if lc["state"] in ACTIVE_STATES else suppressed).append(annotated)
    return {"active": active, "suppressed": suppressed}


def set_state(states=None, key=None, state=None, opts=None):
    """Pure state transition — returns a NEW states map (does not mutate input)."""
    states = states or {}
    opts = opts or {}
    return {
        **states,
        key: {
            "state": state,
            "since": opts.get("now"),
            "snoozeUntil": opts.get("snoozeUntil"),
            "note": opts.get("note"),
        },
    }
